### DISCUSSION TITLE
I would like to ask why texts on exported pdf can not be selected

### DISCUSSION THREAD

**Discussion by SeanXunIC** [mozman/ezdxf#1158] [Q&A] [answered]
Hi! Thank you for taking the time to read my question.
I am trying to export the dxf to pdf, and I find that the **text on pdf cannot be selected** by the cursor of pdf reader (such as Acrobat Pro). I noticed that the texts on pdf files which are exported by AutoCAD can be selected, so the texts can then be processed by other pdf libs such as pymupdf. 
I've tried to understand the source code of ezdxf, and I found that texts may be rendered as a vectorized-graphics format, so the text contents are not stored in pdf. 
I would like to know if I am understanding the code correctly and if there is another way I can save the pdf with both the text content and range information. 
Thank you very much for your attention! Please kindly find my attached pdf export code and the files. 
![image](https://github.com/user-attachments/assets/880bf217-3c92-4398-9333-b7d90a1444c1)
[text export pdf test.zip](https://github.com/user-attachments/files/16838963/text.export.pdf.test.zip)

```
import pathlib
import ezdxf
from ezdxf.addons.drawing import pymupdf, layout
from ezdxf.math import global_bspline_interpolation
from ezdxf.addons.drawing import Frontend, RenderContext
from ezdxf.addons.drawing.config import (
    Configuration,
    BackgroundPolicy,
    ColorPolicy,
    LineweightPolicy,
)
import pymupdf as fitz

# export pdf by ezdxf
def export_pdf(dxf_path, pdf_path, scale=1):
    doc = ezdxf.readfile(dxf_path)
    msp = doc.modelspace()
    output_layout = msp

    ezdxf_page = layout.Page(
                0,  # auto-detect
                0,  # auto-detect
                layout.Units.mm,  # 1 drawing unit = 1mm
                layout.Margins.all(0),
                max_width=1189,  # limit page width to 1189mm
                max_height=841,  # limit page height to 841mm
            )
    settings = layout.Settings(scale=scale)

    backend = pymupdf.PyMuPdfBackend()
    config = Configuration(
            background_policy=BackgroundPolicy.WHITE,
            color_policy=ColorPolicy.COLOR,
            lineweight_policy=LineweightPolicy.RELATIVE,
            lineweight_scaling=1,
        )

    Frontend(RenderContext(doc), backend, config=config).draw_layout(output_layout, ezdxf_page)
    pdf_bytes = backend.get_pdf_bytes(ezdxf_page, settings=settings)
    CWD = pathlib.Path(pdf_path)
    CWD.write_bytes(pdf_bytes)

# read texts by other pdf parse lib such as PyMUPdf
def read_texts_on_pdf(pdf_path):
    doc = fitz.open(pdf_path)
    text_blocks = []
    for page_num in range(len(doc)):
        page = doc[page_num]
        words = page.get_text("word")
        print(words)
    doc.close()


if __name__ == "__main__":
    dxf_path = "test_texts.dxf"
    pdf_path = "output_by_autocad.pdf"
    # export_pdf(dxf_path)
    read_texts_on_pdf(pdf_path)
```

**Comment by mozman** [comment]
You understand correctly, the frontend renders text into filled paths of connected lines and bezier curves using the `fonttools` package. This simplifies the implementation of the backends, produces the same result across all backends, and allows for partial cropping of text. The `drawing` add-on does not currently support any other rendering method.

**Reply by SeanXunIC** [reply]
Thank you for your reply!

**Comment by mbway** [comment]
I have a prototype for passing text data directly to the PDF backend so that it is understood as text and not just shapes. I was able to create a result very close to the result from the 'baked' approach.

There seems to be a problem with rendering `text_fonts.dxf` because the only font that the backend is instructed to load is `DejaVu Sans Condensed`. Even with `--verbose` there is no indication that any fonts are failing to load. So I can't check whether my settings only work for DejaVuSans.

In the following screenshots I have outputted real text in red and the baked text in the default color (white)
![image](https://github.com/user-attachments/assets/af92ddfa-f945-4a02-b304-2e7f3795569c)
![image](https://github.com/user-attachments/assets/1031d132-3efb-434e-bf24-9d39ff326324)

**Reply by mbway** [reply, marked as answer]
see: https://github.com/mozman/ezdxf/pull/1159

**Reply by SeanXunIC** [reply]
I appreciate for the solution you provided, I have successfully generated a PDF with text information. Thank you for taking the time to help with my issue! Although the font information retrieved through pymupdf is incorrect, the fonts display correctly in the PDF editor. So anyway it won't affect my data processing later. 

```
import pathlib
import ezdxf
from ezdxf.addons.drawing import pymupdf, layout
from ezdxf.math import global_bspline_interpolation
from ezdxf.addons.drawing import Frontend, RenderContext
from ezdxf.addons.drawing.config import (
    Configuration,
    BackgroundPolicy,
    ColorPolicy,
    LineweightPolicy,
    TextPolicy
)
import pymupdf as fitz

def read_texts_on_pdf_dicts(pdf_path):
    doc = fitz.open(pdf_path)
    
    for page_num in range(len(doc)):
        page = doc[page_num]
        
        print(f"Page {page_num + 1}:")
        
        text_instances = page.get_text("dict")["blocks"]
        
        for block in text_instances:
            if block["type"] == 0: 
                for line in block["lines"]:
                    for span in line["spans"]:
                        text = span["text"]
                        bbox = span["bbox"]
                        font_size = span["size"]
                        font_name = span["font"]
                        
                        print(f"Text: {text}")
                        print(f"Font: {font_name}, Size: {font_size}")
                        print("Bounding box:")
                        print(f"    Top-left:     ({bbox[0]:.2f}, {bbox[1]:.2f})")
                        print(f"    Bottom-right: ({bbox[2]:.2f}, {bbox[3]:.2f})")
    doc.close()


if __name__ == "__main__":
    # dxf_path = "test_texts.dxf"
    # pdf_path = "output_by_autocad.pdf"
    pdf_path = "./output_NATIVE.pdf"
    # export_pdf(dxf_path, pdf_path)
    read_texts_on_pdf_dicts(pdf_path)
```
The output_NATIVE.pdf is 
[output_NATIVE.pdf](https://github.com/user-attachments/files/16843382/output_NATIVE.pdf)
Part of the print output is:
```
Page 1:
Text: 7300
Font: ArialUnicodeMS, Size: 40.95506286621094
Bounding box:
    Top-left:     (1124.17, 2170.08)
    Bottom-right: (1215.26, 2224.95)
```
![image](https://github.com/user-attachments/assets/94bd590d-60b8-486c-a39f-d8c3a16ad5d9)
